crates/python/qcs_sdk/qpu/experimental.pyi

```diff
@@ -0,0 +1,165 @@
+from typing import Dict, List, Optional, Self, Tuple, final
+
+import numpy as np
+from numpy.typing import NDArray
+
+@final
+class RandomError(ValueError):
+    """
+    An error that may occur while initializing a seed value or
+    generating a QPU PRNG sequence.
+    """
+
+    ...
+
+@final
+class PrngSeedValue:
+    """A seed value for the Rigetti QPU PRNG."""
+
+    def __new__(cls, value: int) -> Self: ...
+
+@final
+class ChooseRandomRealSubRegions:
+    """
+    A class that represents a QPU extern call to pseudo-randomly choose sub-regions of
+    a source memory declaration. The parameters of the call function are as follows:
+
+    * `destination` - The destination memory region to copy the pseudo-randomly chosen
+        sub-region to.
+    * `source` - The source memory region to choose sub-regions from.
+    * `sub_region_size` - The size of each sub-region to choose.
+    * `seed` - A memory reference to an integer value that will be used to seed the PRNG.
+        This value will be mutated to the next element in the PRNG sequence, so you may
+        use it to generate subsequent pseudo-random values.
+
+    Note, `len(destination) % sub_region-size` and `len(source) % sub_region_size` must be 0.
+
+    # Example
+
+    The following example declares a source with 12 real values, representing 4 sub-regions
+    each of size 3. The destination memory region is declared with 3 real values, representing
+    a single sub-region. The seed memory region is declared with a single integer value.
+
+    The `Call` instruction will pseudo-randomly choose a sub-region of the source memory region
+    and copy it to the destination memory region according to the seed value. The seed value is
+    mutated to the next element in the PRNG sequence.
+
+    >>> program = Program()
```

Comment:
I added this doctest as an example because we cannot support a more polished interface (due to the fact that we cannot include `quil-rs` types in the Python interface). Unfortunately, pytest won't actually _run_ this doctest (it doesn't natively support stub files).

One alternative I considered was implementing some high level functions for using `ChooseRandomRealSubRegions` natively in Python by importing `quil-py` types. I stopped short there because there isn't really precedent, however, I think it's probably the best tact.